proof/refine/ARM/Reply_R.thy

```diff
@@ -96,4 +95,487 @@ lemma replyRemoveTCB_st_tcb_at'_sym_ref:
   apply (clarsimp simp: obj_at'_def pred_tcb_at'_def)
   done
 
+lemma getReplyTCB_wp:
+  "\<lbrace>\<lambda>s. reply_at' rptr s \<longrightarrow> obj_at' (\<lambda>reply. P (replyTCB reply) s) rptr s\<rbrace>
+   getReplyTCB rptr
+   \<lbrace>P\<rbrace>"
+  unfolding getReplyTCB_def
+  apply (wpsimp)
+  apply (clarsimp simp: obj_at'_def)
+  done
+
+crunches setReplyTCB
+  for ksReadyQueues[wp]: "\<lambda>s. P (ksReadyQueues s)"
+  and ksSchedulerAction[wp]: "\<lambda>s. P (ksSchedulerAction s)"
+  and valid_queues[wp]: "valid_queues"
+
+crunches getReplyTCB
+  for inv: "P"
+
+lemma setReplyTCB_list_refs_of_replies':
+  "setReplyTCB tcb rptr \<lbrace>\<lambda>s. P (list_refs_of_replies' s)\<rbrace>"
+  unfolding setReplyTCB_def updateReply_def
+  apply wpsimp
+  apply (erule rsubst[where P=P])
+  apply (rule ext)
+  apply (clarsimp simp: list_refs_of_reply'_def obj_at'_def projectKO_eq
+                        list_refs_of_replies'_def opt_map_def
+                 split: option.splits)
+  done
+
+lemma setReply_valid_pde_mappings'[wp]:
+  "setReply p r \<lbrace>valid_pde_mappings'\<rbrace>"
+  unfolding valid_pde_mappings'_def setReply_def
+  apply (wpsimp wp: hoare_vcg_all_lift)
+  done
+
+lemma replyUnlink_valid_pspace'[wp]:
+  "replyUnlink rptr tptr \<lbrace>valid_pspace'\<rbrace>"
```

Comment:
Uh oh, I think there might be some duplicated work here. Victor proved a lot of these lemmas about `replyUnlink` in IpcCancel_R. We need to make sure that they're all consistent before continuing.

Reply:
Yeah, I've removed a lot of now-redundant stuff in `IpcCancel_R`. Wish I'd caught it earlier though.

Reply:
I try to at last skim most PRs that come in, even if I don't have anything to contribute to a review. It helps avoid these sorts of situations